### ISSUE TITLE
performance analysis profiling benchmarking round 1

### ISSUE THREAD

**Issue by Mikolaj** [liqd/aula#649]
More details will be added later on.

Initial IRC transcript:
```
09:05 < mikolaj> let me just summarize here the results of the additional profiling:
09:05 < mikolaj> (nothing groundbreaking nor worrying)
09:06 < mikolaj> I repeated profiling of demo data generation and of tests with on-disk acid-state
09:07 < mikolaj> and also when aula is compiled without optimization (-O0)
09:08 < mikolaj> so, the tests with with -O0; "Finished in 82.0457 seconds" vs 1min; the same memory taken
09:10 < fisx__> that's anti-climactic, but good news i guess.  (:
09:11 < mikolaj> darn, I should build tension first ;)
09:11 < mikolaj> demo data generation with -O0: only a bit slower, 16M vs 8M heap, the memory usage is fast to shrink
09:12 < mikolaj> demo with acid-state on disk and -O0: almost the same time, 16M heap, but memory shrinks only close to the end
09:13 < fisx__> mikolaj: could you open a github issue and put this information there?
09:13 < mikolaj> demo data generation with optimization, but on-disk acid-state: CPU barely used; takes 2 minutes vs immediately
09:14 < mikolaj> heap 8M (was 10M with in-memory), so the longer runtime is IO overhead (15M of disk taken)
09:14 < mikolaj> fisx__: will do
09:14 < mikolaj> BTW, almost finished
09:14 < fisx__> ok (:
09:14  * fisx__ is still listening.
09:14 < fisx__> the 2 minutes worry me, but only a little.
09:15 < fisx__> how many objects are we creating per second?
09:17 < mikolaj> I don't have this data; will repeat
09:17 < mikolaj> BTW, the most CPU-intensive function with on-disk demo data is this :
09:17 < mikolaj> mkRunPersistOnDisk.explainException Persistent.Implementation.AcidState  67.0   13.4   9780 203219080
09:17 < mikolaj> but no errors reported, so it probably just forces here to see if there's exception
09:18 < mikolaj> the log afterwards is not too long and there are no error reported there:
09:18 < mikolaj> -rw-rw-r--  1 mikolaj mikolaj  12K Jun 11 01:16 aulaEventLog.json
09:18 < mikolaj> so, probably nothing wrong; we'd need larger data to make sure
09:18 < mikolaj> though in this case the heap profile looks a bit leaky; but 6M only...
09:18 < mikolaj> so that's it for the profiling data
09:20 < fisx__> mikolaj: one thing you could do is to change the size of the universe in DemoData until creation takes 20 minutes instead of 2.  see if that makes any formerlly insignificant space leaks visible. what do you think?
09:25 < mikolaj_> fisx__: so, I will get the objects/s, try demo data with larger size and put updated results on a separate issue
```

**Comment by Mikolaj** [comment]
Output of -s for the normal on-disk demo data run (heap 8M, 2 minutes taken, mostly IO):

```
   2,309,699,904 bytes allocated in the heap
   1,211,401,576 bytes copied during GC
      11,871,008 bytes maximum residency (165 sample(s))
       1,212,880 bytes maximum slop
              36 MB total memory in use (4 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      4151 colls,     0 par    0.248s   0.467s     0.0001s    0.0022s
  Gen  1       165 colls,     0 par    4.297s   4.371s     0.0265s    0.0487s

  INIT    time    0.000s  (  0.001s elapsed)
  MUT     time    7.187s  ( 56.420s elapsed)
  GC      time    3.650s  (  3.928s elapsed)
  RP      time    0.000s  (  0.000s elapsed)
  PROF    time    0.895s  (  0.910s elapsed)
  EXIT    time    0.000s  (  0.002s elapsed)
  Total   time   11.741s  ( 60.351s elapsed)

  %GC     time      31.1%  (6.5% elapsed)

  Alloc rate    321,379,778 bytes per MUT second

  Productivity  61.3% of total user, 11.9% of total elapsed
```

Top of the -P output:

```
        Mon Jun 13 10:53 2016 Time and Allocation Profiling Report  (Final)

           aula-server +RTS -P -hy -s -RTS

        total time  =       14.91 secs   (14915 ticks @ 1000 us, 1 processor)
        total alloc = 1,500,734,592 bytes  (excludes profiling overheads)

COST CENTRE                         MODULE                              %time %alloc  ticks     bytes

mkRunPersistOnDisk.explainException Persistent.Implementation.AcidState  65.2   13.4   9726 201405384
arbWord                             Arbitrary                             5.1   12.2    759 182964416
userIdeaLocation                    DemoData                              3.8   12.3    571 184088672
putCopy                             Types                                 2.7   10.8    406 162047320
genCommentVote                      DemoData                              2.5    0.5    379   7742632
someOf                              Arbitrary                             1.3    9.9    192 148401544
getCurrentTimestamp                 Action.Implementation                 1.2    1.2    177  17402240
_IdeaLocationSpace                  Types                                 1.0    4.3    154  64030912
putCopy                             Types                                 0.9    2.7    139  41203464
putCopy                             Types                                 0.9    1.8    127  27176520
arbPhraseOf                         Arbitrary                             0.8    4.5    121  67001768
ideaStudentPair                     DemoData                              0.8    1.3    120  19043456
putCopy                             Types                                 0.7    2.3    110  34393528
putCopy                             Types                                 0.7    1.9    106  28094288
putCopy                             Types                                 0.6    1.4     95  20956592
putCopy                             Types                                 0.5    1.5     74  22374232
_ClassSpace                         Types                                 0.5    2.1     70  31301896
putCopy                             Data.Markdown                         0.4    2.2     58  32977744
putCopy                             Types                                 0.3    1.2     40  17444352
```

**Comment by fisx** [comment]
another data point: `DemoData.defaultUniverseSize` counts 8465 objects.  the on-disk run that took 20 seconds thus created 423 objects per second.  so if 4000 users create one object each every 10 seconds, we're fine.

this is a very hand-wavy argument, but i like the conclusion.  (-:

@Mikolaj did I interpret your data correctly?

**Comment by fisx** [comment]
oh, wait, i got it wrong.  let me try again:

another data point: DemoData.defaultUniverseSize counts 8465 objects. the on-disk run that took 120 seconds thus created 70 objects per second.

this is still probably ok, but i'm not sure we shouldn't be able to do better?

**Comment by fisx** [comment]
thanks for the preliminary analysis.  so we have two open questions: (1) what makes creation of objects so slow?  and (2) do we have any slow-acting space leaks?

have you removed the `putStrLn` in `explainException` to see what that does to (1)?

**Comment by fisx** [comment]
(to clarify: we don't care about demo data generation performance as long as it doesn't take hours.  we *do* care about production performance, and if we indeed spend 90% of the time it takes to create an idea for writing some string to stdout, that would be something to be fixed.)

**Comment by Mikolaj** [comment]
I've made more precise measurements, with SSD not too heavily used by other apps, and it's actually only ~36s (versus instant for in-memory), so it's 235 objects/s, which indeed looks acceptable. I've tested and no screen nor log output contributes significantly to this timing. Of these 36s, only ~6s is MUT time and ~3s is GC time, so the rest is most probably just disk operations. I will test some more to see how it scales and if any leaks are apparent.

**Comment by Mikolaj** [comment]
I've increased all numbers in the DemoData.defaultUniverseSize record 10 times and got 30 times longer runtime (17min vs 36s). This is ~80 objects/s so not too bad, especially with heap profiling overhead. This time disk IO still takes, I guess, 20% or 30% of elapsed runtime. The total data on disk at the end is 120M (vs 15M for 10 times smaller data, so that looks linear), but it's apparently rewritten a lot.

But the biggest culprit is GC and profiling overhead due to lots of memory operations. Max heap size on the heap profile is 80M (vs 8M, which is linear, which is probably bad, because acid-state metadata that stays in memory should not increase so much) and the shape is analogous to the one with small dataset and not very alarming, though some of the constructors hint at thunks, which is bad, given that all data should be fully evaluated due to disk storage and whatever is not stored should be forgotten (but is not, due to leaks elsewhere).

That's all I can deduce without code-diving. Generally, for a never optimized code, without even bangs in any record fields, it looks splendid. :)

```
 65,591,791,200 bytes allocated in the heap
 195,696,188,232 bytes copied during GC
     124,158,976 bytes maximum residency (2885 sample(s))
       7,848,792 bytes maximum slop
             312 MB total memory in use (36 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     122155 colls,     0 par    4.179s   8.839s     0.0001s    0.0073s
  Gen  1      2885 colls,     0 par   489.281s  493.831s     0.1712s    0.3681s

  INIT    time    0.000s  (  0.002s elapsed)
  MUT     time  208.705s  (862.292s elapsed)
  GC      time  331.803s  (339.646s elapsed)
  RP      time    0.000s  (  0.000s elapsed)
  PROF    time  161.656s  (163.025s elapsed)
  EXIT    time    0.001s  (  0.035s elapsed)
  Total   time  702.166s  (1201.974s elapsed)

  %GC     time      47.3%  (28.3% elapsed)

  Alloc rate    314,279,276 bytes per MUT second

  Productivity  29.7% of total user, 17.4% of total elapsed
```

The -P output is rather interesting. Worth a look, if we ever work on optimizing the system.

```
        Mon Jun 13 15:18 2016 Time and Allocation Profiling Report  (Final)

           aula-server +RTS -P -hy -s -RTS

        total time  =      266.69 secs   (266694 ticks @ 1000 us, 1 processor)
        total alloc = 40,647,831,408 bytes  (excludes profiling overheads)

COST CENTRE               MODULE                              %time %alloc  ticks     bytes

mkRunPersistOnDisk.opn    Persistent.Implementation.AcidState  36.0    4.9  96028 1990516904
userIdeaLocation          DemoData                             17.5   46.0  46803 18695154752
genCommentVote            DemoData                             13.6    0.2  36267  77425096
_IdeaLocationSpace        Types                                 5.5   16.0  14678 6502662592
relatedStudents           DemoData                              3.0    0.7   8098 276304224
relatedStudents.sameSpace DemoData                              2.9    0.0   7745        16
_ClassSpace               Types                                 2.3    7.7   6265 3146711896
arbWord                   Arbitrary                             2.3    4.4   6156 1794547408
putCopy                   Types                                 1.6    4.1   4270 1673980656
genReply                  DemoData                              1.6    0.1   4154  33690144
==                        Types                                 1.4    0.0   3656        16
==                        Types                                 1.2    0.0   3152         0
_Student                  Types                                 1.1    0.0   2934         0
someOf                    Arbitrary                             0.6    3.6   1609 1455245384
putCopy                   Types                                 0.6    1.0   1482 408269816
arbPhraseOf               Arbitrary                             0.4    1.6   1025 657675664
```

**Comment by Mikolaj** [comment]
From IRC log:
```
21:14 < mikolaj> also, I got this for the first time in tests:
21:14 < mikolaj> daemon [create acid-state checkpoint, archive] timed out after 47m.
21:15 < mikolaj> and so state/ is now 266M, because it's now the new state and the log, I guess; previously it was only the transaction log
21:20 < mikolaj> that's because I left the server for several hours now
21:21 < mikolaj> and actually now the heap profile looks very differently --- the leaks are gone (at least most of them)
21:23 < mikolaj> about half of the memory is reclaimed, which confirms my earlier suspicion we were storing, roughly, 
another copy of the data (probably just as a transaction log in memory, in addtion to state)
21:24 < mikolaj> it turns out only the state checkpoint save evaluates a lot of the data and so also frees the transaction log in memory that was needed to evaluate it
21:25 < mikolaj> fisx__: so I think now we more or less have no permanent leaks
21:25 < mikolaj> that was 1.
21:26 < mikolaj> 2. the cost of disc operations is rather low at the scale of 20min of generating demo data, because if probably fits in the disc caches
21:27 < mikolaj> 3. the cost of GC is high at that scale, because it does not fit in CPU caches any more --- so that's why at this scale GC time becomes higher than disc IO time
21:28 < mikolaj> 4. there's lots of GC because all our data structures are lazy and so we create a lot of closures and evaluate them much later on (when they no more in the cache)
21:30 < mikolaj> 5. so bangs in all datatypes would speed the server up ~2 times in this scenario and at this dataset scale and also reduce memory usage by 2 times
21:31 < mikolaj> 6. at higher scales, with larger datasets the disc IO would probably become the main cost again, so the bangs may not be that useful in the long run
21:32 < mikolaj> fisx__: copying this to the issue; I guess it doesn't make sense to code-dive at this point, unless I'm missing something
```

**Comment by fisx** [comment]
> 21:14 < mikolaj> daemon [create acid-state checkpoint, archive] timed out after 47m.

this just means that a new checkpoint is going to be created.  we implement the loop with the timeout daemon, which issues this log message.  probably a bit misleading in all cases where it is used to implement a loop.

no permanent leaks: great!

all your other points: also great!  (-:

thanks @Mikolaj.  do you still have the heap profiles for later reference?  it would be nice to have them so we can comare them against ghc 8.01 and all-new libraries once we've upgraded.

@andorp @np can we close this?

**Comment by andorp** [comment]
@fisx: Yes

**Comment by Mikolaj** [comment]
@fisx: yes, I will keep the profiles.

**Comment by fisx** [comment]
@Mikolaj i meant: could you upload them here?  alternatively, you could open a sub-folder of docs and dump everything there.  either way i'd like to have the raw data accessible for everybody with access to the code.

**Comment by Mikolaj** [comment]
OK, uploaded the files from the long run, with the state thunks eventually forced and many checkpoint generations visible in the heap profile.

[aula-server.profiles.zip](https://github.com/liqd/aula/files/313530/aula-server.profiles.zip)

**Comment by fisx** [comment]
Ok, thanks again!  I'll close this issue for now with the conclusion that everything is fine performance-wise for the first year.  Yeay!  (-:
